app/src/main/java/nodomain/freeyourgadget/gadgetbridge/util/BengaliLanguageUtils.java

```diff
@@ -1,110 +1,226 @@
+            put("আ", "aa");
+            put("অ", "a");
+            put("ই", "i");
+            put("ঈ", "ii");
+            put("উ", "u");
+            put("ঊ", "uu");
+            put("ঋ", "ri");
+            put("এ", "e");
+            put("ঐ", "oi");
+            put("ও", "o");
+            put("ঔ", "ou");
+            put("ক", "k");
+            put("খ", "kh");
+            put("গ", "g");
+            put("ঘ", "gh");
+            put("ঙ", "ng");
+            put("চ", "ch");
+            put("ছ", "chh");
+            put("জ", "j");
+            put("ঝ", "jh");
+            put("ঞ", "Ng");
+            put("ট", "T");
+            put("ঠ", "Th");
+            put("ড", "D");
+            put("ঢ", "Dh");
+            put("ণ", "N");
+            put("ত", "t");
+            put("থ", "th");
+            put("দ", "d");
+            put("ধ", "dh");
+            put("ন", "n");
+            put("প", "p");
+            put("ফ", "ph");
+            put("ব", "b");
+            put("ভ", "bh");
+            put("ম", "m");
+            put("য", "J");
+            put("র", "r");
+            put("ল", "l");
+            put("শ", "sh");
+            put("ষ", "Sh");
+            put("স", "s");
+            put("হ", "h");
+            put("ড়", "rh");
+            put("ঢ়", "rH");
+            put("য়", "y");
+            put("ৎ", "t");
+            put("০", "0");
+            put("১", "1");
+            put("২", "2");
+            put("৩", "3");
+            put("৪", "4");
+            put("৫", "5");
+            put("৬", "6");
+            put("৭", "7");
+            put("৮", "8");
+            put("৯", "9");
+            put("া", "aa");
+            put("ি", "i");
+            put("ী", "ii");
+            put("ু", "u");
+            put("ূ", "uu");
+            put("ৃ", "r");
+            put("ে", "e");
+            put("ো", "o");
+            put("ৈ", "oi");
+            put("ৗ", "ou");
+            put("ৌ", "ou");
+            put("ং", "ng");
+            put("ঃ", "h");
+            put("ঁ", "nN");
+            put("।", ".");
         }
     };
 
-    private final static HashMap<Character, String> symbols = new HashMap<Character, String>() {
-        {
-            put('ব', "w");
-            put('য়', "y");
-        }
-    };
+    // The regex to extract Bengali characters in nested groups.
+    private final static String pattern = "(র্){0,1}(([অ-হড়-য়])(্([অ-মশ-হড়-য়]))*)((‍){0,1}(্([য-ল]))){0,1}([া-ৌ]){0,1}|([্ঁঃংৎ০-৯।])| ";
+    private final static Pattern bengaliRegex = Pattern.compile(pattern);
 
-    private final static HashMap<Character, String> joins = new HashMap<Character, String>() {
-        {
-            put('৳', "$");
+    private static String getVal(String key) {
+        if (key != null) {
+            boolean hasKey = composites.containsKey(key);
+            if (hasKey) {
+                return composites.get(key);
+            }
+            hasKey = letters.containsKey(key);
+            if (hasKey) {
```

Comment:
AFAICS, containsKey() should not be necessary here. You could just use get() and check for a `null` result. Similarly below.

Reply:
Oh, It gives a Exception only in Hashtable then?

Reply:
Hello, did the way you said. I was actually learning Java back then had some error related null key. Now I don't code that bad. :smile: